modules/patching/diff.go

```diff
@@ -0,0 +1,275 @@
+package patching
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Diff represents a single change in the document.
+type Diff struct {
+	Insertion  bool
+	StartIndex int
+	Changes    string
+}
+
+// NewDiff creates a new diff object from the given parameters
+func NewDiff(insertion bool, startIndex int, changes string) *Diff {
+	return &Diff{
+		Insertion:  insertion,
+		StartIndex: startIndex,
+		Changes:    changes,
+	}
+}
+
+// NewDiffFromString parses a diff from its string representation.
+func NewDiffFromString(str string) (*Diff, error) {
+	regex, err := regexp.Compile("\\d+:(\\+|-)\\d+:.+")
+	if err != nil {
+		return nil, err
+	}
+	if !regex.MatchString(str) {
+		return nil, errors.New("Illegal patch format; should be %d:+%d:%s or %d:-%d:%s")
+	}
+
+	parts := strings.Split(str, ":")
+	diff := Diff{}
+
+	// Parse startIndex
+	diff.StartIndex, err = strconv.Atoi(parts[0])
+	if err != nil {
+		return nil, fmt.Errorf("Invalid offset: %s", parts[0])
+	}
+
+	// Switch on diff type
+	switch parts[1][0] {
+	case '+':
+		diff.Insertion = true
+	case '-':
+		diff.Insertion = false
+	default:
+		return nil, fmt.Errorf("Invalid operation: %s", parts[1][0:1])
```

Comment:
I'd recommend you declare the types of these errors so that we can type them in future tests